File: PythonAlgorithmInterView/선형자료구조/자신을제외한배열의곱.py

#rock 문제 풀이방법. time limit exceeded가 걸린다.
def productExceptSelf(self, nums: list[int]) -> list[int]:

    result = []

    p = 1
    for i in range(len(nums)):
        result.append(p)
        p *= nums[i]

    p = 1
    for i in range(len(nums), 0, -1):
        result[i-1] *= p
        p *= nums[i-1]

    return result

if __name__ == '__main__':
    nums = [1, 2, 3, 4]
    # nums = [-1, 1, 0, -3, 3]

    print(productExceptSelf(None, nums))

File: PythonAlgorithmInterView/선형자료구조/자신을제외한배열의곱.js
/**
 * @param {number[]} nums
 * @return {number[]}
 * 238번 문제.
 */
 var productExceptSelf = function(nums) {
    
    let result = [];
    let numsSize = nums.length;

    let pointer = 1;

    for(let i = 0; i< numsSize; i++){
        result.push(pointer)
        pointer *= nums[i]
    }

    pointer = 1;

    for(let i = numsSize-1; i >= 0; i--){
        result[i] *= pointer
        pointer *= nums[i]
    }

    return result

};

let nums = [1,2,3,4];
console.log(productExceptSelf(nums))